primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1033 @@
+	fn from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Self;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Returns `None` if `d == 0` or `n / d` exceeds accuracy.
+	fn checked_from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Option<Self>;
+
+	/// Checked multiplication for integer type `N`.
+	fn checked_mul_int<N: FixedPointOperand>(self, other: N) -> Option<N>;
+
+	/// Saturating multiplication for integer type `N`.
+	fn saturating_mul_int<N: FixedPointOperand>(self, other: N) -> N;
+
+	/// Checked division for integer type `N`.
+	fn checked_div_int<N: FixedPointOperand>(self, other: N) -> Option<N>;
+
+	/// Saturating division for integer type `N`.
+	fn saturating_div_int<N: FixedPointOperand>(self, other: N) -> N;
+
+	/// Performs a saturated multiplication and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	fn saturating_mul_int_acc<N: FixedPointOperand>(self, int: N) -> N;
+
+	/// Saturating absolute value. Returning MAX if `parts == Inner::MIN` instead of overflowing.
+	fn saturating_abs(self) -> Self;
+
+	/// Takes the reciprocal (inverse), `1 / self`.
+	fn reciprocal(self) -> Option<Self> {
+		Self::one().checked_div(&self)
+	}
+
+	/// Returns zero.
+	fn zero() -> Self;
+
+	/// Checks if the number is zero.
+	fn is_zero(&self) -> bool;
+
+	/// Returns one.
+	fn one() -> Self;
+
+	/// Checks if the number is positive.
+	fn is_positive(self) -> bool;
+
+	/// Checks if the number is negative.
+	fn is_negative(self) -> bool;
+
+	/// Returns the integer part.
+	fn trunc(self) -> Self;
+
+	/// Returns the fractional part.
+	fn frac(self) -> Self;
+
+	/// Returns the smallest integer greater than or equal to a number.
+	fn ceil(self) -> Self;
+
+	/// Returns the largest integer less than or equal to a number.
+	fn floor(self) -> Self;
+
+	/// Returns the nearest integer to a number. Round half-way cases away from 0.0.
+	fn round(self) -> Self;
+
+	/// Computes the absolute value of self.
+	fn abs(self) -> Self;	
+}
+
+struct I129 {
+	value: u128,
+	negative: bool,
+}
+
+impl<N: FixedPointOperand> From<N> for I129 {
+	fn from(n: N) -> I129 {
+		if n < N::zero() {
+			let value: u128 = n.checked_neg()
+				.map(|n| n.unique_saturated_into())
+				.unwrap_or(N::max_value().unique_saturated_into().saturating_add(1));
+			I129 { value, negative: true }
+		} else {
+			I129 { value: n.unique_saturated_into(), negative: false }
+		}
+	}
+}
+
+fn from_i129<N: FixedPointOperand>(n: I129) -> Option<N> {
+	let max_plus_one: u128 = N::max_value().unique_saturated_into().saturating_add(1);
+	if n.value == max_plus_one && n.negative {
+		Some(N::min_value())
+	} else {
+		let unsigned_inner: N = n.value.try_into().ok()?;
+		let inner = if n.negative { unsigned_inner.checked_neg()? } else { unsigned_inner };
+		Some(inner)
+	}
+}
+
+fn to_bound<N: FixedPointOperand, D: FixedPointOperand, R: Bounded>(n: N, m: D) -> R {
```

Comment:
Could you add a small comment here about what this function does. Not 100% clear to me tbh

Reply:
Sure. When we need to saturate this function should choose the `min` or `max` bound depending on the sign of the operands.